web/app/mu-plugins/filters.php

```diff
@@ -97,17 +101,96 @@ function fix_core_resource_urls( $url ) {
  *
  * is_main_site will return true if the site is not multisite.
  *
+ * @since 1.1.0
  * @param string $url The URL to check.
- *
  * @return string The filtered URL.
  */
-function adjust_main_site_urls( $url ) {
+function adjust_main_site_urls( string $url ) : string {
 	// If this is the main site, drop the /wp.
-	if ( is_main_site() ) {
+	if ( is_main_site() && ! __is_login_url( $url ) ) {
 		$url = str_replace( '/wp/', '/', $url );
 	}
 
 	return $url;
 }
-add_filter( 'home_url', 'adjust_main_site_urls', 9 );
-add_filter( 'site_url', 'adjust_main_site_urls', 9 );
+add_filter( 'home_url', __NAMESPACE__ . '\\adjust_main_site_urls', 9 );
+add_filter( 'site_url', __NAMESPACE__ . '\\adjust_main_site_urls', 9 );
+
+/**
+ * Add /wp prefix to all admin and login URLs.
+ * Since /wp is where the core files are installed, this normalizes all non-front-facing urls to use the correct url structure.
+ *
+ * @since 1.1.0
+ * @param string $url The URL to check.
+ * @return string The corrected admin or login URL (or the base url if not an admin or login url).
+ */
+function add_wp_prefix_to_login_and_admin_urls( string $url ) : string {
+	if ( __is_login_url( $url ) ) {
+		if ( strpos( $url, '/wp/' ) === false ) {
+			$url = __normalize_wp_url( preg_replace( '/(\/wp-(login|admin))/', '/wp/$1', $url ) );
+		}
+	}
+
+	return $url;
+}
```

Comment:
can you guess what I'm going to say?
```suggestion
function add_wp_prefix_to_login_and_admin_urls( string $url ) : string {
	if (  ! __is_login_url( $url ) ) {
		return $url;
	}
	if ( strpos( $url, '/wp/' ) !== false ) {
		return $url;
	}
	return __normalize_wp_url( preg_replace( '/(\/wp-(login|admin))/', '/wp/$1', $url ) );
}
```